src/operations.jl

```diff
@@ -1,7 +1,18 @@
+"""
+    ArrayReference
+
+A type for encoding an array reference `A[i,j]` occurring inside an `@avx` block.
+
+# Fields
 
+$(TYPEDFIELDS)
+"""
 struct ArrayReference
+    "The array variable"
     array::Symbol
+    "The list of indices (e.g., `[:i, :j]`), or `name(op)` for computed indices."
     indices::Vector{Symbol}
+    "Index offset, e.g., `a[i+7]` would store the `7`. Only used for small (`Int8`) offsets."
     offsets::Vector{Int8}
```

Comment:
To add, `offsets` is used currently to identify loops that may benefit from unrolling. For example, in the exponential smoothing example, where we have both `X[Ipre,i,Ipost]` and `X[Ipre,i-1,Ipost]`. Unrolling `i` lets us save on `X` reloads.

Later, we can use it to identify loop-carried dependencies.

It is possible someone writes a loop using `X[Ipre,i + 130,Ipost]` and `X[Ipre,i + 129,Ipost]`, but before we reply on it for loop-carried dependencies (where it's just an optimization instead of correctness issue), I think we can live without that. It would be more common for people writing loops to define the `i` loop to incorporate that `129` offset.
Maybe when finally adding loop carried dependency support, we should increase it to `Int16`, and use an `Int128` in `condense_loopsets` (to hold the 8 offsets).
Or switch to something more flexible.

Reply:
The "save reloads" is awesome. Such an impressive package!

I suspect `Int8` is big enough for the *vast* majority of stencils. I just wanted to document the restriction, but I don't think the restriction is likely to get in the way of practical applications.

Reply:
[This](https://github.com/chriselrod/LoopVectorization.jl/blob/07de1e49655bcae5cb1634416868bfe77d85240f/src/determinestrategy.jl#L461) is the part of `determinestrategy.jl` that looks for offsets.

Lowering also currently special cases when `tiled` is the loop that has the offset. LLVM will probably handle when `unrolled` is that loop, but I'll probably add special case handling to be sure.

5 sets of benchmarks (this should be on your CartesianIndexing compiler pass PR), setup
```julia
using LoopVectorization, BenchmarkTools, Test
function smoothdim!(s, x, α, Rpre, irng::AbstractUnitRange, Rpost)
    ifirst, ilast = first(irng), last(irng)
    ifirst > ilast && return s
    @inbounds @fastmath for Ipost in Rpost
        # Initialize the first value along the filtered dimension
        for Ipre in Rpre
            s[Ipre, ifirst, Ipost] = x[Ipre, ifirst, Ipost]
        end
        # Handle all other entries
        for i = ifirst+1:ilast
            for Ipre in Rpre
                s[Ipre, i, Ipost] = α*x[Ipre, i, Ipost] + (1-α)*x[Ipre, i-1, Ipost]
            end
        end
    end
    s
end
function smoothdim_avx!(s, x, α, Rpre, irng::AbstractUnitRange, Rpost)
    ifirst, ilast = first(irng), last(irng)
    ifirst > ilast && return s
    @avx for Ipost in Rpost
         for Ipre in Rpre
            s[Ipre, ifirst, Ipost] = x[Ipre, ifirst, Ipost]
            for i = ifirst+1:ilast
                s[Ipre, i, Ipost] = α*x[Ipre, i, Ipost] + (1-α)*x[Ipre, i-1, Ipost]
            end
        end
    end
    s
end
function smoothdim_ifelse_avx!(s, x, α, Rpre, irng::AbstractUnitRange, Rpost)
    ifirst, ilast = first(irng), last(irng)
    ifirst > ilast && return s
    @avx for Ipost in Rpost, i = ifirst:ilast, Ipre in Rpre
        xi = x[Ipre, i, Ipost]
        xim = i > ifirst ? x[Ipre, i-1, Ipost] : xi
        s[Ipre, i, Ipost] = α*xi + (1-α)*xim
    end
    s
end

M = 8;
x = rand(M,M,M,M,M);
dest1, dest2, dest3 = similar(x), similar(x), similar(x);
α = 0.3
```
Note that `M = 8`. With `M = 11`, this benchmark is mostly bandwidth limited, meaning `smoothdim_avx!` performs worst (except when the Base version shows excess allocations at `d == 1` and `d == 2`, as shown below), because `LoopVectorization` splits the loop into two separate passes.
For larger problems (like `M = 11`), given that it's bandwidth limited, cache-based optimizations would help more.

```julia
julia> d = 1
1

julia> Rpre  = CartesianIndices(axes(x)[1:d-1]);

julia> Rpost = CartesianIndices(axes(x)[d+1:end]);

julia> @benchmark smoothdim!($dest1, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  19.69 MiB
  allocs estimate:  598016
  --------------
  minimum time:     30.222 ms (0.00% GC)
  median time:      30.363 ms (0.00% GC)
  mean time:        31.355 ms (3.19% GC)
  maximum time:     34.358 ms (0.00% GC)
  --------------
  samples:          160
  evals/sample:     1

julia> @benchmark smoothdim_avx!($dest2, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     19.985 μs (0.00% GC)
  median time:      20.483 μs (0.00% GC)
  mean time:        20.514 μs (0.00% GC)
  maximum time:     44.879 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark smoothdim_ifelse_avx!($dest3, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     13.287 μs (0.00% GC)
  median time:      13.619 μs (0.00% GC)
  mean time:        13.645 μs (0.00% GC)
  maximum time:     52.820 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @test dest1 ≈ dest2 ≈ dest3
Test Passed
```
These excess allocations were the case on regular Julia master as well. Any idea where the allocations are coming from? `@code_warntype` looks fine, as far as I can tell.

When `M == 8`, LoopVectorization is at it's slowest when `d == 1`, because `d == 1` does not allow loads to be eliminated. That's because when `d == 1`, the `i` loop is vectorized
```julia
x1 = X[i - 1,...]
x2 = X[i    ,...]
x3 = X[i + W - 1,...]
x4 = X[i + W    ,...]
```
`x2 == x3` will only be true if `W = 1`, where `W` is the increment between subsequent loop iterations (and also the width of SIMD vectors).

When `M == 11`, this was actually the fastest loop, since LoopVectorization would traverse this loop nest in the most cache-friendly minor. Load-elimination doesn't help when memory-bandwidth-bottlenecked, because the memory of the load you're eliminating had to already be available for loading.

`d == 2` shows the same `Base` problem, but demonstrates the benefit of load elimination when the arrays are smaller:
```julia
julia> d = 2
2

julia> Rpre  = CartesianIndices(axes(x)[1:d-1]);

julia> Rpost = CartesianIndices(axes(x)[d+1:end]);

julia> @benchmark smoothdim!($dest1, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  19.69 MiB
  allocs estimate:  598016
  --------------
  minimum time:     30.178 ms (0.00% GC)
  median time:      30.279 ms (0.00% GC)
  mean time:        31.293 ms (3.15% GC)
  maximum time:     34.247 ms (0.00% GC)
  --------------
  samples:          160
  evals/sample:     1

julia> @benchmark smoothdim_avx!($dest2, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     8.142 μs (0.00% GC)
  median time:      8.339 μs (0.00% GC)
  mean time:        8.361 μs (0.00% GC)
  maximum time:     21.851 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     3

julia> @benchmark smoothdim_ifelse_avx!($dest3, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     9.478 μs (0.00% GC)
  median time:      9.849 μs (0.00% GC)
  mean time:        9.874 μs (0.00% GC)
  maximum time:     50.904 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @test dest1 ≈ dest2 ≈ dest3
Test Passed
```

The standard Julia version speeds up 1000-fold once the allocations are elliminated:
However, starting with `d == 3`:
```julia
julia> d = 3;

julia> Rpre  = CartesianIndices(axes(x)[1:d-1]);

julia> Rpost = CartesianIndices(axes(x)[d+1:end]);

julia> @benchmark smoothdim!($dest1, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     33.337 μs (0.00% GC)
  median time:      33.525 μs (0.00% GC)
  mean time:        33.587 μs (0.00% GC)
  maximum time:     59.428 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark smoothdim_avx!($dest2, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     7.335 μs (0.00% GC)
  median time:      7.569 μs (0.00% GC)
  mean time:        7.584 μs (0.00% GC)
  maximum time:     20.087 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     4

julia> @benchmark smoothdim_ifelse_avx!($dest3, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     10.104 μs (0.00% GC)
  median time:      10.498 μs (0.00% GC)
  mean time:        10.527 μs (0.00% GC)
  maximum time:     56.877 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @test dest1 ≈ dest2 ≈ dest3
Test Passed
```
With `d == 4`:
```julia
julia> d = 4;

julia> Rpre  = CartesianIndices(axes(x)[1:d-1]);

julia> Rpost = CartesianIndices(axes(x)[d+1:end]);

julia> @benchmark smoothdim!($dest1, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     32.990 μs (0.00% GC)
  median time:      33.333 μs (0.00% GC)
  mean time:        33.383 μs (0.00% GC)
  maximum time:     61.634 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark smoothdim_avx!($dest2, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     8.977 μs (0.00% GC)
  median time:      9.302 μs (0.00% GC)
  mean time:        9.332 μs (0.00% GC)
  maximum time:     78.718 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark smoothdim_ifelse_avx!($dest3, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     14.053 μs (0.00% GC)
  median time:      14.669 μs (0.00% GC)
  mean time:        14.700 μs (0.00% GC)
  maximum time:     52.863 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @test dest1 ≈ dest2 ≈ dest3
Test Passed
```
and finally, `d == 5`:
```julia
julia> d = 5;

julia> Rpre  = CartesianIndices(axes(x)[1:d-1]);

julia> Rpost = CartesianIndices(axes(x)[d+1:end]);

julia> @benchmark smoothdim!($dest1, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     33.267 μs (0.00% GC)
  median time:      33.487 μs (0.00% GC)
  mean time:        33.537 μs (0.00% GC)
  maximum time:     51.475 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark smoothdim_avx!($dest2, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     9.378 μs (0.00% GC)
  median time:      9.693 μs (0.00% GC)
  mean time:        9.727 μs (0.00% GC)
  maximum time:     81.404 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark smoothdim_ifelse_avx!($dest3, $x, $α, $Rpre, axes($x, $d), $Rpost)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     8.809 μs (0.00% GC)
  median time:      9.071 μs (0.00% GC)
  mean time:        9.087 μs (0.00% GC)
  maximum time:     22.969 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     3

julia> @test dest1 ≈ dest2 ≈ dest3
Test Passed
```

How large is `M` in practice?
If it tends to be much larger, iterating over statically sized tiles may help.

Static sizing would also help the Base Julia version, which wont benefit from SIMD otherwise unless the inner most loop is of length `size(x,1) >= 4*W`, where W is the SIMD vector width.